params/include/alps/params/serialize_variant.hpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 1998-2017 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+
+/** @file serialize_variant.hpp
+    
+    @brief Header for boost::variant serialization
+
+    @todo FIXME: Move to utilities
+*/
+
+#ifndef ALPS_PARAMS_SERIALIZE_VARIANT_HPP_d116b6a1742b418f851e0dbb87004644
+#define ALPS_PARAMS_SERIALIZE_VARIANT_HPP_d116b6a1742b418f851e0dbb87004644
+
+#include <boost/variant.hpp>
+#include <boost/mpl/for_each.hpp>
+#include <stdexcept>
+#include <boost/optional.hpp>
+#include <boost/lexical_cast.hpp> // for error messages
+#include <boost/type_index.hpp> // for error messages
+
+namespace alps {
+    namespace detail {
+
+        /// Class to mem-serialize and mem-deserialize a `boost::variant` over MPL type sequence MPLSEQ
+        template <typename MPLSEQ, typename CONSUMER, typename PRODUCER>
+        class variant_serializer {
+          public:
+            typedef typename boost::make_variant_over<MPLSEQ>::type variant_type;
+            typedef CONSUMER consumer_type;
+            typedef PRODUCER producer_type;
+
+          private:
+            /// Visitor to call consumer on the bound value
+            struct consume_visitor : public boost::static_visitor<> {
+                consumer_type& consumer_;
+
+                consume_visitor(consumer_type& consumer) : consumer_(consumer) {}
+                
+                template <typename T>
+                void operator()(const T& val) const
+                {
+                    consumer_(val);
+                }
+            };
+
+            /// Functor class to reconstruct object via producer
+            struct maker {
+                producer_type& producer_;
+                variant_type& var_;
+
+                maker(producer_type& producer, variant_type& var)
+                    : producer_(producer), var_(var)
+                { }
+
+                // Always assigns --- therefore, applies last successful outcome!
+                template <typename T>
+                void operator()(const T&) {
+                    boost::optional<T> maybe_val=producer_((T*)0);
+                    if (maybe_val) var_=*maybe_val;
+                }
+            };
+
+          public:
+            
+            /// Send the variant to a consumer
+            static void consume(consumer_type& consumer, const variant_type& var)
+            {
+                boost::apply_visitor(consume_visitor(consumer), var);
+            }
+            
+            /// Receive the variant from a producer
+            static variant_type produce(producer_type& producer)
+            {
+                variant_type var;
+                boost::mpl::for_each<MPLSEQ>(maker(producer, var));
```

Comment:
Do you mean, something along the lines of:
```c++
boost::variant produce(const void* data, int tag) {
     if (tag==tags<int>::value)  return make<int>(data);
else if (tag==tags<double>::value)  return make<double>(data);
else if (tag==tags<dblvec>::value)  return make<dblvec>(data);
//.....
else throw runtime_error("Unknown tag");
}
```
(or using `switch`/`case`) ?

The idea behind the current code is to repeatedly ask the producer "is it type int? is it type double?" and so on, for all supported types. It is up to the transport-specific producer to figure it out: for MPI transport it is based on the value of an integer tag `which`, for HDF5 it is based on the type of the dataset, and so on. The decision "is it type T?" is deliberately delegated to the producer.

Suppose, for the sake of example, the producer uses an integer tag to determine the type; then the current code is *almost* analogous to `producer()` essentially making the comparison inside the call; something along the lines of:
```c++
template <typename T>
optional<T> producer(const void* data, int tag) {
   if (tag==tags<T>::value) return make<T>(data);
   else return boost::none;
}

boost::variant produce(const void* data, int tag) {
 optional maybe_val;
 maybe_val=producer<int>(data,tag); if (maybe_val) return *maybe_val;
 maybe_val=producer<double>(data,tag); if (maybe_val) return *maybe_val;
 maybe_val=producer<dblvec>(data,tag); if (maybe_val) return *maybe_val;
 // ....
 throw runtime_error("Unknown tag");
}
```
Yes, it has an extra 20 comparisons and extra function calls (i believe they will be inlined anyway). But we are not concerned with efficiency here (as of now, at least), and the `mpl::for_each` loop is compact and easy to write and understand.

To be fair, the example above will pick up the first match, making 10 comparisons on average (and even less, as most of the parameters likely have type int, not vector of strings); the code as it is now will pick the last match, making all 20 comparisons. This i may be able to optimize, if it's worth it.

Reply:
Yes, I meant a switch statement, which is usually compiled into a jump table.

Alright, I agree that performance is non-issue here, so it's basically my OCD that protests against doing O(N) operations for solving an intrinsically O(1) problem.

Reply:
I share the sentiment and I believe it is possible to cast the problem into a `switch/case` form (e.g., let the transport-specific producer return an integer tag value that goes into `switch`). But then changing the list of supported types will require changing the deserialization code too, unless we enlist Boost preprocessor tricks. It is a performance vs convenience trade-off. It can be thought through, but maybe it's of a lower priority. Thank you for bringing it up, though!